### PR TITLE
Add Unit tests for ComicInfo and ComicBookInfo formats

### PR DESCRIPTION
**Pull request by carif**
Add unit tests for the ComicInfo and ComicBookInfo format using xunit and NSubstitute. While adding the unit tests, I found a few bugs that I fixed as well.

The test data for the  ComicBookInfo unit test is taken from the documentation of the format: https://code.google.com/archive/p/comicbookinfo/wikis/Example.wiki
The test data for the ComicInfo unit test is real data from an actual manga. There is one exception to this: The image size of the pages is an arbitrary chosen value instead of using the real image size (stored in bytes).

In either case, I provided **more test data than is actually tested against**. This should allow developers who do not own comics / mangas to contribute by testing their changes against the test data. This could  **serve as a reference to add missing fields to the Jellyfin Server** to allow the plugin to save the entire metadata to the server and then let clients use those additional metadata fields for their own purposes (e.g. reading direction for manga).

No unit tests for the epub format as I ran out of time. :/

### CODE REVIEW

File: Jellyfin.Plugin.Bookshelf/Providers/ComicBookInfo/ComicBookInfoProvider.cs [hunk 1]
```diff
@@ -168,16 +159,51 @@
...
+        {
+            try
+            {
+                return new CultureInfo(language).DisplayName;
+            }
+            catch (Exception)
+            {
+                // Ignored
+                return null;
+            }
```
[the remark this answers is not in the document]
Comment (carif) [reply Jellyfin.Plugin.Bookshelf/Providers/ComicBookInfo/ComicBookInfoProvider.cs:204]: Same CodeQL warning as before as the methods name got changed while the code remained the same.

File: Jellyfin.Plugin.Bookshelf/Providers/ComicBookInfo/ComicBookInfoProvider.cs [hunk 2]
```diff
@@ -178,6 +169,41 @@ private MetadataResult<Book> SaveMetadata(ComicBookInfoFormat comic)
...
+                metadataResult.AddPerson(personInfo);
+            }
+        }
+
+        /// <inheritdoc />
+        public string? ReadCultureInfoInto(string language)
+        {
+            try
+            {
+                return new CultureInfo(language).DisplayName;
```
Comment (cvium) [inline Jellyfin.Plugin.Bookshelf/Providers/ComicBookInfo/ComicBookInfoProvider.cs:198]: Use the static method `CultureInfo.GetCultureInfo` instead. It caches internally.

File: tests/Jellyfin.Plugin.Bookshelf.Tests/ComicBookInfoProviderTest.cs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+    private readonly ComicBookInfoProvider _uut;
+
+    public ComicBookInfoProviderTest()
+    {
+        _comicBookInfoFormat = GenerateTestData();
+
+        var fileSystem = Substitute.For<IFileSystem>();
+        var logger = Substitute.For<ILogger<ComicBookInfoProvider>>();
+        _uut = new ComicBookInfoProvider(fileSystem, logger);
```
Comment (cvium) [inline tests/Jellyfin.Plugin.Bookshelf.Tests/ComicBookInfoProviderTest.cs:24]: What does uut stand for? I'm familiar with `System Under Test` (SUT), but not uut

**Review by cvium — APPROVED** [review]

File: tests/Jellyfin.Plugin.Bookshelf.Tests/ComicBookInfoProviderTest.cs (hunk 3 above)
Comment (carif) [reply tests/Jellyfin.Plugin.Bookshelf.Tests/ComicBookInfoProviderTest.cs:24]: `Unit Under Test`
Not sure which one is used more often, but I'm familiar with both terms.

File: tests/Jellyfin.Plugin.Bookshelf.Tests/Jellyfin.Plugin.Bookshelf.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    <PackageReference Include="NSubstitute.Analyzers.CSharp" Version="1.0.15">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="xunit" Version="2.4.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="coverlet.collector" Version="3.1.0">
```
Comment (ReenigneArcher) [inline tests/Jellyfin.Plugin.Bookshelf.Tests/Jellyfin.Plugin.Bookshelf.Tests.csproj:24]:
```suggestion
    <PackageReference Include="xunit" Version="2.5.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.5.1">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="coverlet.collector" Version="6.0.0">
```

**Review by ReenigneArcher — COMMENTED** [review]
Saw a comment in Discord about wanting a review. Mostly just looking for ideas for a PR (https://github.com/LizardByte/Themerr-jellyfin/pull/101) I have on one of my plugins to add unit testing.

Noticed some out of date dependencies.

Maybe add an ability to log messages? From my very limited experience with `xunit` and `dotnet test`, the standard output from `dotnet test` is not helpful at all. I created a simple [logger](https://github.com/LizardByte/Themerr-jellyfin/blob/da1c190fd0d5e54b41d70aed726ffb16567bf5a4/Jellyfin.Plugin.Themerr.Tests/TestLogger.cs#L6) to use during testing. Then run with tests with `dotnet test --logger "console;verbosity=detailed"` to actually get some valuable output.

**Comment by carif** [comment]
@cvium @ReenigneArcher Thx for the review. :)

> Maybe add an ability to log messages? From my very limited experience with `xunit` and `dotnet test`, the standard output from `dotnet test` is not helpful at all. I created a simple [logger](https://github.com/LizardByte/Themerr-jellyfin/blob/da1c190fd0d5e54b41d70aed726ffb16567bf5a4/Jellyfin.Plugin.Themerr.Tests/TestLogger.cs#L6) to use during testing. Then run with tests with `dotnet test --logger "console;verbosity=detailed"` to actually get some valuable output.

I'm not sure we need logging for those simple tests, especially since the classes under test don't log anything (in the methods under test) anyway. Should we want to log something in those tests, xunit provides a way to do so without writing a custom log class: https://xunit.net/docs/capturing-output

**Pull request merged**